File: AgentLens/Services/HomeAssistant/HomeAssistantWebhookID.swift
import Foundation

// MARK: - Webhook ID generator
//
// HA webhook IDs are public-ish strings (anyone with the ID can fire
// the webhook on the LAN), so they must be high-entropy. We follow
// HA's own approach: 32 random URL-safe characters drawn from the
// system CSPRNG. We prefix with `openburnbar_cast_recover_` so that a
// glance at HA's automations.yaml makes the source obvious.

enum HomeAssistantWebhookID {

    static let prefix = "openburnbar_cast_recover_"
    static let secretLength = 32

    /// Cryptographically random URL-safe webhook id.
    /// Deterministic in tests via `randomBytes`.
    static func generate(randomBytes: () -> [UInt8] = defaultRandomBytes) -> String {
        let bytes = randomBytes()
        // Base32-ish alphabet (lowercase a-z + 0-9), no padding, URL safe.
        let alphabet = Array("abcdefghijklmnopqrstuvwxyz0123456789")
        var output = ""
        output.reserveCapacity(secretLength)
        for byte in bytes.prefix(secretLength) {
            output.append(alphabet[Int(byte) % alphabet.count])
        }
        return prefix + output
    }

    /// Returns true if the supplied id was minted by us.
    static func isOurs(_ id: String) -> Bool {
        id.hasPrefix(prefix) && id.count == prefix.count + secretLength
    }

    /// Default RNG path — `SystemRandomNumberGenerator`.
    static func defaultRandomBytes() -> [UInt8] {
        var rng = SystemRandomNumberGenerator()
        var output: [UInt8] = []
        output.reserveCapacity(secretLength)
        for _ in 0..<secretLength {
            output.append(UInt8(rng.next() % 256))
        }
        return output
    }
}

File: windows/integrations/homeassistant/OpenBurnBar.Integrations.HomeAssistant/HomeAssistantWebhookId.cs
using System;
using System.Security.Cryptography;
using System.Text;

namespace OpenBurnBar.Integrations.HomeAssistant;

// High-entropy HA webhook id generator.
//
// Parity: AgentLens/Services/HomeAssistant/HomeAssistantWebhookID.swift
//   enum HomeAssistantWebhookID (prefix / secretLength / generate / isOurs).
//
// HA webhook ids are public-ish LAN strings, so they must be high-entropy. We
// follow HA's own approach: 32 random chars from a lowercase base36 alphabet,
// prefixed so the source is obvious in the user's automations.yaml.

public static class HomeAssistantWebhookId
{
    public const string Prefix = "openburnbar_cast_recover_";
    public const int SecretLength = 32;

    private static readonly char[] Alphabet = "abcdefghijklmnopqrstuvwxyz0123456789".ToCharArray();

    /// Cryptographically random URL-safe webhook id. `randomBytes` is injectable
    /// so tests can drive it deterministically (parity with the Swift default arg).
    public static string Generate(Func<byte[]>? randomBytes = null)
    {
        var bytes = (randomBytes ?? DefaultRandomBytes)();
        var output = new StringBuilder(Prefix.Length + SecretLength);
        output.Append(Prefix);
        var count = Math.Min(SecretLength, bytes.Length);
        for (var i = 0; i < count; i++)
        {
            output.Append(Alphabet[bytes[i] % Alphabet.Length]);
        }
        return output.ToString();
    }

    /// True when the supplied id was minted by us (prefix + exact length).
    public static bool IsOurs(string id) =>
        id.StartsWith(Prefix, StringComparison.Ordinal) &&
        id.Length == Prefix.Length + SecretLength;

    /// Default CSPRNG path.
    public static byte[] DefaultRandomBytes()
    {
        var output = new byte[SecretLength];
        RandomNumberGenerator.Fill(output);
        return output;
    }
}
